compiler/src/iree/compiler/Codegen/Dialect/GPU/Transforms/ConcretizeMmaShapes.cpp

```diff
@@ -0,0 +1,115 @@
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Arith/Utils/Utils.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Linalg/IR/LinalgInterfaces.h"
+#include "mlir/Dialect/Linalg/Transforms/Transforms.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/SCF/Transforms/TileUsingInterface.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/OpDefinition.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Interfaces/FunctionInterfaces.h"
+#include "mlir/Interfaces/LoopLikeInterface.h"
+#include "mlir/Support/LogicalResult.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir::iree_compiler::IREE::GPU {
+
+#define GEN_PASS_DEF_CONCRETIZEMMASHAPESPASS
+#include "iree/compiler/Codegen/Dialect/GPU/Transforms/Passes.h.inc"
+
+namespace {
+struct ConcretizeMmaShapesPass final
+    : impl::ConcretizeMmaShapesPassBase<ConcretizeMmaShapesPass> {
+  void runOnOperation() override;
+};
+} // namespace
+
+struct ConcretizeMmaShapes final : OpRewritePattern<IREE::GPU::MultiMmaOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(IREE::GPU::MultiMmaOp mmaOp,
+                                PatternRewriter &rewriter) const override {
+    IREE::GPU::MmaInterfaceAttr kind = mmaOp.getKind();
+    SmallVector<ReassociationIndices> lhsReassociations, rhsReassociations,
+        accReassociations;
+    RankedTensorType lhsConcreteType, rhsConcreteType, accConcreteType;
+    if (failed(kind.materializeOperandConcreteShape(
+            rewriter, IREE::GPU::MMAFragment::Lhs, mmaOp.getLhs(),
+            mmaOp.getLhsPermutation(), lhsReassociations, lhsConcreteType))) {
+      return failure();
+    }
+    if (failed(kind.materializeOperandConcreteShape(
+            rewriter, IREE::GPU::MMAFragment::Rhs, mmaOp.getRhs(),
+            mmaOp.getRhsPermutation(), rhsReassociations, rhsConcreteType))) {
+      return failure();
+    }
+    if (failed(kind.materializeOperandConcreteShape(
+            rewriter, IREE::GPU::MMAFragment::Acc, mmaOp.getAcc(),
+            mmaOp.getAccPermutation(), accReassociations, accConcreteType))) {
+      return failure();
+    }
+    Location loc = mmaOp->getLoc();
+    Value concreteLhs =
+        rewriter
+            .create<tensor::ExpandShapeOp>(loc, lhsConcreteType, mmaOp.getLhs(),
+                                           lhsReassociations)
+            .getResult();
+    Value concreteRhs =
+        rewriter
+            .create<tensor::ExpandShapeOp>(loc, rhsConcreteType, mmaOp.getRhs(),
+                                           rhsReassociations)
+            .getResult();
+    Value concreteAcc =
+        rewriter
+            .create<tensor::ExpandShapeOp>(loc, accConcreteType, mmaOp.getAcc(),
+                                           accReassociations)
+            .getResult();
```

Comment:
Rather than making this a single pass, I would split it up into two places. One that runs immediately after (or during) PackToIntrinsics for just the accumulator, and then another that happens immediately before subgroup tiling for the input operands. Then tiling won't have to reason about reshapes at all.